lib/Dialect/ESI/ESIPasses.cpp

```diff
@@ -1388,6 +1388,169 @@ void ESIPortsPass::updateInstance(HWModuleExternOp mod, InstanceOp inst) {
   inst.erase();
 }
 
+//===----------------------------------------------------------------------===//
+// Lower high-level ESI types to HW conversions and pass.
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// Lower all "high-level" ESI types on modules to some lower construct.
+struct ESITypesPass : public LowerESITypesBase<ESITypesPass> {
+  void runOnOperation() override;
+
+private:
+  void lowerMod(HWMutableModuleLike mod);
+  void lowerInstance(HWInstanceLike inst);
+
+  /// Lower an individual port, modifing 'port'. Returns 'true' if the port was
+  /// changed.
+  bool lowerPort(HWMutableModuleLike mod, hw::PortInfo &port);
+
+  /// Cache the lowered types to avoid running the same type lowering again and
+  /// again.
+  DenseMap<WindowType, hw::UnionType> typeCache;
+  /// Track the modules we've modified both to make lookup easier and to avoid
+  /// attempting to lower instances which do not have to be lowered.
+  DenseMap<StringAttr, hw::HWMutableModuleLike> modsMutated;
+};
+} // anonymous namespace
+
+bool ESITypesPass::lowerPort(HWMutableModuleLike mod, hw::PortInfo &port) {
+  auto window = hw::type_dyn_cast<WindowType>(port.type);
+  if (!window)
+    return false;
+  hw::UnionType &lowered = typeCache[window];
+  if (!lowered)
+    lowered = window.getLoweredType();
+  port.type = lowered;
+  return true;
+}
+
+void ESITypesPass::lowerMod(HWMutableModuleLike mod) {
```

Comment:
hmm... I know this is probably the "CIRCT-canonical" way of doing module-level type modifications (mainly due to `mod.modifyPorts`). I'm wondering whether you've also (now or at an earlier point) experimented with using the conversion infrastructure?
It seems applicable since you're basically DIYing a type converter + symbol cache + type conversion materialization.